### PR TITLE
Add pagination to review list

### LINKED ISSUE #1267: Reviews tab: Add pagination to review list

### PR DESCRIPTION
**Pull request by ctarda**
Fixes #1267 

Implement pagination in the review list, using the exact same pattern as other listings, like Orders. Which, by the way, being at least the third view controller where we use the same pattern, suggests that there might be an abstraction here that we might want to extract: a paginated table view. 😉 

ezgif-2-180c8ef41279

## Changes
- Aggregate a SyncingCoordinator, and implement its delegate from ReviewsViewController
- Pass the syncing coordinator down to the tableview delegate on willDisplayCell, so that we can decide when to load the net page. 
- An update to the logic that persists Reviews to coredata in the ProductsReviewStore. With new pages, previous data was getting removed 😬 

## Testing
- As with other PRs related to Reviews, testing is a bit cubmersome (courtesy of some of the storyboards):
- Checkout the branch, run the tests
- Undo the feature flag. Because the NotificationsViewController is instantiated in a Storyboard, it will be necessary to touch `FeatureFlag.swift` to:
```
        case .reviews:
            return BuildConfiguration.current == .localDeveloper
```
and it will also be necessary to change the class of the first view controller in the Notifications.storyboard from NotificationsViewController to ReviewsViewController

- Navigate to the reviews tab, on a site with more than 25 reviews. Scroll down and check if the reviews list is paginated

Update release notes:

- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Notifications/ReviewsViewModel.swift [hunk 1]
```diff
@@ -103,19 +109,22 @@ extension ReviewsViewModel {
...
         }
 
-        let action = ProductReviewAction.synchronizeProductReviews(siteID: siteID, pageNumber: 1, pageSize: 25) { error in
+        let action = ProductReviewAction.synchronizeProductReviews(siteID: siteID, pageNumber: pageNumber, pageSize: pageSize) { error in
             if let error = error {
                 DDLogError("⛔️ Error synchronizing reviews: \(error)")
                 ServiceLocator.analytics.track(.reviewsListLoadFailed,
                                                withError: error)
             } else {
+                let loadingMore = pageNumber != 1
```
Comment (jaclync) [inline WooCommerce/Classes/ViewRelated/Notifications/ReviewsViewModel.swift]: nit: how about using `Settings.firstPage` here?

File: WooCommerce/Classes/ViewRelated/Notifications/DefaultReviewsDataSource.swift [hunk 2]
```diff
@@ -224,6 +231,16 @@ extension DefaultReviewsDataSource: ReviewsInteractionDelegate {
...
 
+//    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
+//
+//        // Preserve the Cell Height
+//        // Why: Because Autosizing Cells, upon reload, will need to be laid yout yet again. This might cause
+//        // UI glitches / unwanted animations. By preserving it, *then* the estimated will be extremely close to
+//        // the actual value. AKA no flicker!
+//        //
+//        estimatedRowHeights[indexPath] = cell.frame.height
+//    }
```
Comment (jaclync) [inline WooCommerce/Classes/ViewRelated/Notifications/DefaultReviewsDataSource.swift]: do we still need the code here?

**Comment by ctarda** [comment]
👋🏼 @jaclync @mindgraffiti . I just pushed an update that should fix the issue when switching stores. 

> I felt pretty stuck scrolling the product reviews on my iPhone 6s Plus device (iOS 12)

What do you mean exactly? Poor scroll performance?

**Comment by jaclync** [comment]
Nice investigation and fix, I can now see Product reviews after switching stores! A side issue that might be related to the site ID somewhere not refreshed after switching stores: 

- Go to Reviews tab --> each review row says "{Someone} left a review on {a product}" without a product name
- Tap on a review --> the product row shown on the top shows the product name and a link to open the product url
- Switch to another store
- Go to Reviews tab --> **each review row says "{Someone} left a review on" without a product name**
- Tap on a review **--> the product row shown on the now shows "unknown" and the link doesn't work. The products are** 

The products seem loaded on the Products tab, so I'm guessing it's the site ID in the product predicate?

store 1 | store 2
-- | --
IMG_7631 | IMG_7630

**Comment by jaclync** [comment]
> I felt pretty stuck scrolling the product reviews on my iPhone 6s Plus device (iOS 12)
What do you mean exactly? Poor scrolling performance?

It feels like dropping frames much more than scrolling on the Orders and Products tab on my iPhone 6s Plus device. Have you tried it on a device? Not sure if it's noticeable on a newer device (my primary iPhone XR is on iOS 13.2 and I couldn't build it from Xcode 😞). My guess is the String+HTML on the snippet for each cell. Before the Product Reviews tab, we didn't show many reviews so I rarely had to scroll. Now that there could be more than one page of reviews, I started scrolling and noticed the scrolling performance

**Comment by ctarda** [comment]
I'm sorry @jaclync I am not sure how I did not notice that, but I should have. Sorry...😊 

I've just pushed another update that should refresh the products results controller after switching stores. 

One more thing I have noticed. The tests do not seem to pass on CI, but do pass locally for me. 

With regards to the performance, it seems to scroll reasonably well on my test device (a 7s+). Shall we defer a proper profiling to a separate issue?

**Comment by ctarda** [comment]
Interesting, there is [one test failing on CI](https://circleci.com/gh/woocommerce/woocommerce-ios/3119?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) that passes locally on my machine

**Comment by ctarda** [comment]
The weird thing about that test is that it started failing (very) recently. I have modified it for now in order to make the build pass, it would be worth looking into it in depth. I'll add a ticket for it.

In the meantime... merging!

**Pull request merged**